File: conversions/is-odd/elide-is-odd.rb
#!/usr/bin/env ruby

# Ruby Integration Example for elide-is-odd
#
# This demonstrates calling the TypeScript is-odd implementation
# from Ruby using Elide's polyglot capabilities.
#
# Benefits:
# - One is-odd implementation shared across TypeScript and Ruby
# - Consistent odd/even logic across microservices
# - No Ruby .odd? method inconsistencies
# - Perfect for Rails validations, helpers

puts "=== Ruby Calling TypeScript is-odd ===\n"

# NOTE: Exact syntax depends on Elide's Ruby polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Ruby support is ready

# Assuming Elide provides something like:
# is_odd_module = Elide.require('./elide-is-odd.ts')

# Example 1: ActiveRecord Validation
# class User < ApplicationRecord
#   validate :user_id_must_be_odd
#
#   private
#
#   def user_id_must_be_odd
#     unless is_odd_module.default(id)
#       errors.add(:id, "must be an odd number")
#     end
#   end
# end

# Example 2: Array Filtering
# numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
# odd_numbers = numbers.select { |n| is_odd_module.default(n) }
# puts "Odd numbers: #{odd_numbers}"
# # => [1, 3, 5, 7, 9]

# Example 3: Rails Helper
# module ApplicationHelper
#   def highlight_odd_rows(index)
#     is_odd_module.default(index) ? 'bg-gray-100' : 'bg-white'
#   end
# end
#
# # In view:
# # <% @items.each_with_index do |item, i| %>
# #   <tr class="<%= highlight_odd_rows(i) %>">
# #     <td><%= item.name %></td>
# #   </tr>
# # <% end %>

# Example 4: Sidekiq Job Batching
# class DataProcessor
#   include Sidekiq::Worker
#
#   def perform(batch_id)
#     # Process differently based on odd/even batch
#     strategy = is_odd_module.default(batch_id) ? :fast : :thorough
#     process_with_strategy(strategy)
#   end
#
#   def process_with_strategy(strategy)
#     # Implementation
#   end
# end

# Example 5: Pagination
# class PaginationHelper
#   def self.page_class(page_num)
#     if is_odd_module.default(page_num)
#       'odd-page'
#     else
#       'even-page'
#     end
#   end
# end
#
# (1..10).each do |page|
#   puts "Page #{page}: #{PaginationHelper.page_class(page)}"
# end

# Example 6: Rake Task
# namespace :users do
#   desc "Process users with odd IDs"
#   task :process_odd => :environment do
#     User.find_each do |user|
#       if is_odd_module.default(user.id)
#         puts "Processing user ##{user.id}"
#         # Process user
#       end
#     end
#   end
# end

puts "Real-world use cases:"
puts "- ActiveRecord validations"
puts "- View helper methods (row highlighting)"
puts "- Background job batching strategies"
puts "- Pagination UI classes"
puts "- Data filtering and selection"
puts "- Algorithm implementations"
puts ""

puts "Example: Rails E-Commerce Platform"
puts "┌──────────────────────────────────────┐"
puts "│   Elide is-odd (TypeScript)         │"
puts "│   elide-is-odd.ts                   │"
puts "└──────────────────────────────────────┘"
puts "         ↓           ↓           ↓"
puts "    ┌────────┐  ┌────────┐  ┌────────┐"
puts "    │ Node.js│  │  Rails │  │ Sidekiq│"
puts "    │  API   │  │  App   │  │ Worker │"
puts "    └────────┘  └────────┘  └────────┘"
puts "         ↓           ↓           ↓"
puts "    Same validation everywhere!"
puts ""

puts "Problem Solved:"
puts "Before: Ruby .odd? + JS % operator = potential inconsistencies"
puts "After: One Elide implementation = consistent odd/even checks"
puts ""

puts "Rails Integration (when ready):"
puts "  # config/initializers/is_odd.rb"
puts "  IS_ODD = Elide.require('./elide-is-odd.ts')"
puts "  "
puts "  # app/helpers/application_helper.rb"
puts "  module ApplicationHelper"
puts "    def odd?(num)"
puts "      IS_ODD.default(num)"
puts "    end"
puts "  end"
puts ""

puts "Edge Cases Handled:"
puts "- Negative numbers: odd?(-3) -> true"
puts "- Zero: odd?(0) -> false"
puts "- Non-integers: odd?(3.5) -> false"
puts "- Strings: odd?('5') -> true"

File: conversions/is-odd/elide-is-odd.py
#!/usr/bin/env python3
"""
Python Integration Example for elide-is-odd

This demonstrates calling the TypeScript is-odd implementation from Python
using Elide's polyglot capabilities.

Benefits:
- One is-odd implementation shared across TypeScript and Python
- Consistent odd number checking across services
- No Python custom function needed
- Perfect for data validation pipelines
"""

# NOTE: Exact syntax depends on Elide's Python polyglot API (currently alpha)
# This is a CONCEPTUAL example - adjust when Python support is ready

# Assuming Elide provides something like:
# from elide import require
# is_odd_module = require('./elide-is-odd.ts')

print("=== Python Calling TypeScript is-odd ===\n")

# Example 1: Basic odd number checking
# print("Basic checks:")
# print(f"is_odd(3): {is_odd_module.default(3)}")    # True
# print(f"is_odd(2): {is_odd_module.default(2)}")    # False
# print(f"is_odd(7): {is_odd_module.default(7)}")    # True
# print(f"is_odd(-1): {is_odd_module.default(-1)}")  # True
# print()

# Example 2: Filter odd numbers from list
# numbers = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
# odd_numbers = [n for n in numbers if is_odd_module.default(n)]
# print(f"Odd numbers from {numbers}:")
# print(f"  {odd_numbers}")
# print()

# Example 3: Data validation pipeline
# def validate_lottery_number(num):
#     """Validate lottery numbers (must be odd)"""
#     if not is_odd_module.default(num):
#         raise ValueError(f"Lottery number {num} must be odd")
#     return num
#
# try:
#     valid_num = validate_lottery_number(7)
#     print(f"Valid lottery number: {valid_num}")
#     validate_lottery_number(4)  # Will raise error
# except ValueError as e:
#     print(f"Validation error: {e}")
# print()

# Example 4: Batch processing
# data_records = [
#     {'id': 1, 'value': 10},
#     {'id': 2, 'value': 20},
#     {'id': 3, 'value': 30},
#     {'id': 4, 'value': 40},
#     {'id': 5, 'value': 50}
# ]
#
# odd_id_records = [r for r in data_records if is_odd_module.default(r['id'])]
# print("Records with odd IDs:")
# for record in odd_id_records:
#     print(f"  ID {record['id']}: value={record['value']}")
# print()

# Example 5: Alternating patterns
# def generate_pattern(length):
#     """Generate alternating pattern based on odd/even index"""
#     pattern = []
#     for i in range(length):
#         if is_odd_module.default(i):
#             pattern.append('X')
#         else:
#             pattern.append('O')
#     return ''.join(pattern)
#
# pattern = generate_pattern(10)
# print(f"Pattern (O=even index, X=odd index): {pattern}")
# print()

print("Real-world use case:")
print("- Python data validation checks if numbers are odd")
print("- Uses same TypeScript implementation as Node.js API")
print("- Guarantees consistent odd/even logic across entire stack")
print("- No need for custom Python odd checking")
print()

print("Example: Data Processing Pipeline")
print("┌─────────────────────────────────────┐")
print("│   Elide is-odd (TypeScript)        │")
print("│   conversions/is-odd/elide-is-odd.ts│")
print("└─────────────────────────────────────┘")
print("         ↓                   ↓")
print("    ┌────────┐          ┌────────┐")
print("    │ Node.js│          │ Python │")
print("    │  API   │          │Pipeline│")
print("    └────────┘          └────────┘")
print("         ↓                   ↓")
print("    Same odd checking logic!")
print()

print("Problem Solved:")
print("Before: Python (n % 2 == 1) + JavaScript custom logic = potential inconsistencies")
print("After: One Elide implementation = 100% consistent odd checking")
print()

print("Performance Benefits:")
print("- Zero overhead with Elide")
print("- Instant odd checking")
print("- Shared runtime across languages")
print()

print("When Elide Python API is ready, usage will be:")
print("  from elide import require")
print("  is_odd = require('./elide-is-odd.ts')")
print("  result = is_odd.default(3)  # That's it!")
